parser_library/src/lexing/logical_line.h

```diff
@@ -132,19 +136,47 @@ struct logical_line
         friend bool operator==(const const_iterator& a, const const_iterator& b) noexcept
         {
             assert(a.m_logical_line == b.m_logical_line);
-            return a.m_segment == b.m_segment && a.m_col == b.m_col;
+            return a.m_segment_it == b.m_segment_it && a.m_col_it == b.m_col_it;
         }
         friend bool operator!=(const const_iterator& a, const const_iterator& b) noexcept { return !(a == b); }
 
         bool same_line(const const_iterator& o) const noexcept
         {
             assert(m_logical_line == o.m_logical_line);
-            return m_segment == o.m_segment;
+            return m_segment_it == o.m_segment_it;
+        }
+
+        const_iterator transform_into(const logical_line& other_ll) const noexcept
```

Comment:
I think I am starting to see what is the overall direction and use case here... so what about
```
std::pair<size_t, size_t> get_coordinates() const noexcept
{
    assert(m_col != std::string_view::const_iterator()); // valid only for non-end iterators
    return { m_segment->code_offset + std::distance(m_segment.code.begin(), m_col ), std::distance(m_logical_line->segments.begin(), m_segment)};
}
```
given how our range works, it should probably be code_offset_utf16 + utf16 distance (but I wouldn't bother with this at the moment).

Range between two iterators is then
```
range text_range(logical_line::const_iterator b, logical_line::const_iterator e) noexcept
{
    assert("b != ll.end()");
    const auto [bx, by] = b.get_coordinates();
    const position b_pos(bx, by);
    if (b == e) // empty range
        return range(b_pos, b_pos);
    else
        return range(b_pos, "std::prev(e).get_coordinates() + 1 column (really next character size in utf-16 )");
}
```

Reply:
Seems to work perfectly! Nonetheless, where would you put `range text_range(...) noexcept` definition?

I have it in `db2_logical_line_helper` so far but maybe you've had more general location in mind like `range_provider`?

Reply:
range_provider would be a good candidate, but I am fine with keeping it in the db2_logical_line_helper, or just a freestanding function in the db2 preprocessor or range provider source file.